secp256k1-sys/vendor-libsecp.sh

```diff
@@ -1,42 +1,96 @@
+DEFAULT_DEPEND_DIR="$SECP_SYS/depend"
+DEFAULT_SECP_REPO=https://github.com/bitcoin-core/secp256k1.git
+
+: "${SECP_VENDOR_VERSION_CODE:=$DEFAULT_VERSION_CODE}"
+: "${SECP_VENDOR_DEPEND_DIR:=$DEFAULT_DEPEND_DIR}"
+: "${SECP_VENDOR_SECP_REPO:=$DEFAULT_SECP_REPO}"
+# CP_NOT_CLONE lets us just copy a directory rather than git cloning.
+# This is usually a bad idea, since it will bring in build artifacts or any other
+# junk from the source directory, but may be useful during development or CI.
+: "${SECP_VENDOR_CP_NOT_CLONE:=no}"
+
+echo "Using version code $SECP_VENDOR_VERSION_CODE. Set SECP_VENDOR_VERSION_CODE to override."
+echo "Using depend directory $SECP_VENDOR_DEPEND_DIR. Set SECP_VENDOR_DEPEND_DIR to override."
+echo "Using secp repository $SECP_VENDOR_SECP_REPO. Set SECP_VENDOR_SECP_REPO to override."
 
-PARENT_DIR=$1
-VERSIONCODE=$2
-REV=$3
-DIR=secp256k1
-ORIGDIR=$(pwd)
-
-while true; do
-    read -r -p "$PARENT_DIR/$DIR will be deleted [yn]: " yn
-    case $yn in
-        [Yy]* ) break;;
-        [Nn]* ) exit;;
-        * ) echo "Please answer yes or no.";;
+# Parse command-line options
+SECP_REV=""
+FORCE=no
+while (( "$#" )); do
+    case "$1" in
+    -f)
+        FORCE=yes
+        ;;
+    *)
+        if [ -z "$SECP_REV" ]; then
+            echo "Using secp256k1 revision $SECP_REV."
+            SECP_REV="$1"
+        else
+            echo "WARNING: ignoring unknown command-line argument $1"
+        fi
+        ;;
     esac
+    shift
 done
 
-cd "$PARENT_DIR" || exit 1
-rm -rf "$DIR"
-git clone https://github.com/bitcoin-core/secp256k1.git "$DIR"
-cd "$DIR"
-if [ -n "$REV" ]; then
-    git checkout "$REV"
+if [ -z "$SECP_REV" ]; then
+    echo "WARNING: No secp256k1 revision specified. Will use whatever we find at the git repo."
+fi
+echo
+
+# Check if we will do anything destructive.
+
+if [ "$FORCE" == "no" ]; then
+    if ! git diff --quiet -- "*.rs"; then
+        echo "ERROR: There appear to be modified source files. Check these in or pass -f (some source files will be modified to have symbols renamed)."
+        exit 2
+    fi
+    if ! git diff --quiet -- "$SECP_VENDOR_DEPEND_DIR"; then
+        echo "ERROR: The depend directory appears to be modified. Check it in or pass -f (this directory will be deleted)."
+        exit 2
+    fi
 fi
-HEAD=$(git rev-parse HEAD)
-cd ..
-echo "# This file was automatically created by $0" > ./secp256k1-HEAD-revision.txt
-echo "$HEAD" >> ./secp256k1-HEAD-revision.txt
 
-# We need to make some source changes to the files.
+DIR=./secp256k1
+
+pushd "$SECP_VENDOR_DEPEND_DIR" > /dev/null
+rm -rf "$DIR" || true
+
+# Clone the repo. As a special case, if the repo is a local path and we have
+# not specified a revision, just copy the directory rather than using 'git clone'.
+# This lets us use non-git repos or dirty source trees as secp sources.
+if [ "$SECP_VENDOR_CP_NOT_CLONE" == "yes" ]; then
+    cp -r "$SECP_VENDOR_SECP_REPO" "$DIR"
```

Comment:
Might be worth excluding `.git` somehow?

Reply:
I was mainly concerned about speed of copying - large repos can make a significant difference. If you're not going to avoid copying in the first place I guess don't bother.